realsense/scene_perception/win/scene_perception_object.cc

```diff
@@ -493,108 +478,18 @@ void ScenePerceptionObject::OnRunPipeline() {
-    state_ = IDLE;
-    return;
-  }
-
-  // Keep color and depth images.
-  // Allocate memory for images if necessary.
-  if (!latest_color_image_) {
-    PXCImage::ImageInfo image_info;
-    memset(&image_info, 0, sizeof(image_info));
-    image_info.width = sample->color->QueryInfo().width;
-    image_info.height = sample->color->QueryInfo().height;
-    image_info.format = PXCImage::PIXEL_FORMAT_RGB32;
-    latest_color_image_ = session_->CreateImage(&image_info);
-  }
-
-  if (!latest_depth_image_) {
-    PXCImage::ImageInfo image_info;
-    memset(&image_info, 0, sizeof(image_info));
-    image_info.width = sample->depth->QueryInfo().width;
-    image_info.height = sample->depth->QueryInfo().height;
-    image_info.format = PXCImage::PIXEL_FORMAT_DEPTH;
-    latest_depth_image_ = session_->CreateImage(&image_info);
-  }
-
-  // Copy the images.
-  latest_color_image_->CopyImage(sample->color);
-  latest_depth_image_->CopyImage(sample->depth);
-
-  // Get the depth quality.
-  float quality = 0.0;
-  quality = scene_perception_->CheckSceneQuality(sample);
-
-  if ((state_ == INITIALIZED) && checking_event_on_) {
-    CheckingEvent event;
-    event.quality = quality;
-    scoped_ptr<base::ListValue> eventData(new base::ListValue);
-    eventData->Append(event.ToValue().release());
-
-    DispatchEvent("checking", eventData.Pass());
-  }
-
-  if (state_ == STARTED) {
-    PXCScenePerception::TrackingAccuracy accuracy;
-    accuracy = scene_perception_->QueryTrackingAccuracy();
-    float pose[12];
-    scene_perception_->GetCameraPose(pose);
-
-    if (sampleprocessed_event_on_) {
-      SampleProcessedEvent event;
-      event.quality = quality;
-      event.accuracy = toJsAccuracy(accuracy);
-      for (int i = 0; i < 12; ++i) {
-        event.camera_pose.push_back(pose[i]);
-      }
-      scoped_ptr<base::ListValue> eventData(new base::ListValue);
-      eventData->Append(event.ToValue().release());
-
-      DispatchEvent("sampleprocessed", eventData.Pass());
-    }
-
-    if (meshupdated_event_on_
-        && !doing_meshing_updating_
-        && scene_perception_->IsReconstructionUpdated()) {
-      DispatchEvent("meshupdated");
-    }
-
-    // Keep vertices from current camera pose.
-    // PXCPoint3DF32 * internalWidth * internalHeight.
-    // PXCPoint3DF32 = [float x, float y, float z]
-    if (!latest_vertices_) {
-      latest_vertices_.reset(new uint8[3 * sizeof(float) *
-          sp_intrinsics_.imageSize.width * sp_intrinsics_.imageSize.height]);
-    }
-    scene_perception_->GetVertices(
-        reinterpret_cast<PXCPoint3DF32*>(latest_vertices_.get()));
-
-    // Keep normals from current camera pose.
-    if (!latest_normals_) {
-      latest_normals_.reset(new uint8[3 * sizeof(float) *
-          sp_intrinsics_.imageSize.width * sp_intrinsics_.imageSize.height]);
-    }
-    scene_perception_->GetNormals(
-        reinterpret_cast<PXCPoint3DF32*>(latest_normals_.get()));
-  }
-
+  command_queue_.ExecutePendingCommands();
```

Comment:
The existing implementation of `OnRunPipeline` executes:

1. calls `AcquireFrame(true)` feed a frame (say frame0) to SP module.
1. query and cache the data of frame0 from SP module
1. dispatch `sampleprocessed` event of frame0 to JS
1. calls `ReleaseFrame`
1. schedule the next run of `OnRunPipeline`

So when JS calls `getXXXData`, if the JS is able to keep the 30 FPS (so the cached data is not updated with frame1, the JS is able to get the data of frame0. And by using one cache instead of a queue, it limits the memory usage when JS is slow.

Now in this PR, let's say frame0 is processed, when working on frame1:

1. calls `AcquireFrame(true)` feed a frame (say frame1) to SP module.
1. executes pending commands (of frame0), it includes dispatching `sampleprocessed` event to JS
1. calls `ReleaseFrame`
1. schedule the next run of `OnRunPipeline`

So when JS calls `getXXXData` on frame0 event, the implementation will postpone these commands when SP module processed frame1 and return the data from frame1.

I see this is the major difference of the existing implementation and this PR.

Is my understanding correct?